subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/VirtualComponentMetadataResolver.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts.ivyservice.resolveengine;
+
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.internal.component.external.model.VirtualComponentIdentifier;
+import org.gradle.internal.component.model.ComponentOverrideMetadata;
+import org.gradle.internal.resolve.resolver.ComponentMetaDataResolver;
+import org.gradle.internal.resolve.result.BuildableComponentResolveResult;
+
+/**
+ * A component metadata resolver which should be used first in chain of resolvers as it will make sure
+ * that we never "find" a virtual component, by shortcutting resolution if we find the marker interface
+ * for virtual components.
+ */
+class VirtualComponentMetadataResolver implements ComponentMetaDataResolver {
```

Comment:
To me, this is much better than the 'discovery' of virtual components.

Now that we know up-front if a component is virtual or not, I'd like to revert the changes made in #5957. If a platform is not virtual, then we should fail if it does not resolve. I suspect that we've left the fallback code in place to construct a virtual platform if we are told (via `belongsTo("", false)` that a platform is published and it turns out to be missing.

The `retryIfMissing` boolean is the sort of complexity that might seem harmless enough, but could cause us pain further down the line.